File: Demo2/app/src/main/java/com/example/demo/adapter/PaginatedAdapter.java
package com.example.demo.adapter;

import android.app.Activity;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public abstract class PaginatedAdapter<ITEM, VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<VH> {
    private final List<ITEM> mDataSet = new ArrayList<>();
    private OnPaginationListener mListener;
    private int mCurrentPage = 1;
    private RecyclerView mRecyclerView;
    private boolean loadingNewItems = true;


    @NonNull
    public abstract VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    public abstract void onBindViewHolder(@NonNull VH holder, int position);

    public int getItemCount() {
        return mDataSet.size();
    }

    public void submitItems(Collection<? extends ITEM> collection) {
        mDataSet.addAll(collection);
        if (mListener != null) {
            mListener.onCurrentPage(mCurrentPage);
            int mPageSize = 20;
            if (collection.size() == mPageSize) {
                loadingNewItems = false;
            } else {
                mListener.onFinish();
            }
        }
    }

    public int getStartPage() {
        return 1;
    }

    public void setDefaultRecyclerView(Activity activity, int recyclerViewId) {
        RecyclerView recyclerView = activity.findViewById(recyclerViewId);
        recyclerView.setLayoutManager(new LinearLayoutManager(activity));
        recyclerView.setHasFixedSize(true);
        this.mRecyclerView = recyclerView;
        initPaginating();
        setAdapter();
    }

    private void setAdapter() {
        mRecyclerView.setAdapter(this);
    }


    private void initPaginating() {
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
                if (layoutManager != null) {
                    int totalItemCount = layoutManager.getItemCount();
                    int lastVisible = layoutManager.findLastVisibleItemPosition();
                    boolean endHasBeenReached = lastVisible + 2 >= totalItemCount;
                    if (totalItemCount > 0 && endHasBeenReached) {
                        if (mListener != null) {
                            if (!loadingNewItems) {
                                loadingNewItems = true;
                                mListener.onNextPage(++mCurrentPage);
                            }
                        }
                    }
                }
            }
        });
    }

    public void setOnPaginationListener(OnPaginationListener onPaginationListener) {
        this.mListener = onPaginationListener;
    }

    public interface OnPaginationListener {
        void onCurrentPage(int page);

        void onNextPage(int page);

        void onFinish();
    }
}

File: Demo2/app/src/main/java/com/example/demo/adapter/PaginatedAdapter.kt
package com.example.demo.adapter

import androidx.recyclerview.widget.RecyclerView
import com.example.demo.adapter.PaginatedAdapter.OnPaginationListener
import android.view.ViewGroup
import android.app.Activity
import androidx.recyclerview.widget.LinearLayoutManager
import java.util.ArrayList

abstract class PaginatedAdapter<ITEM, VH : RecyclerView.ViewHolder?> : RecyclerView.Adapter<VH>() {
    private val mDataSet: MutableList<ITEM> = ArrayList()
    private var mListener: OnPaginationListener? = null
    private var mCurrentPage = 1
    private var mRecyclerView: RecyclerView? = null
    private var loadingNewItems = true
    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH
    abstract override fun onBindViewHolder(holder: VH, position: Int)
    override fun getItemCount(): Int {
        return mDataSet.size
    }

    fun submitItems(collection: Collection<ITEM>) {
        mDataSet.addAll(collection)
        if (mListener != null) {
            mListener!!.onCurrentPage(mCurrentPage)
            val mPageSize = 20
            if (collection.size == mPageSize) {
                loadingNewItems = false
            } else {
                mListener!!.onFinish()
            }
        }
    }

    val startPage: Int
        get() = 1

    fun setDefaultRecyclerView(activity: Activity, recyclerViewId: Int) {
        val recyclerView: RecyclerView = activity.findViewById(recyclerViewId)
        recyclerView.layoutManager = LinearLayoutManager(activity)
        recyclerView.setHasFixedSize(true)
        mRecyclerView = recyclerView
        initPaginating()
        setAdapter()
    }

    private fun setAdapter() {
        mRecyclerView!!.adapter = this
    }

    private fun initPaginating() {
        mRecyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val layoutManager = recyclerView.layoutManager as LinearLayoutManager?
                if (layoutManager != null) {
                    val totalItemCount = layoutManager.itemCount
                    val lastVisible = layoutManager.findLastVisibleItemPosition()
                    val endHasBeenReached = lastVisible + 2 >= totalItemCount
                    if (totalItemCount > 0 && endHasBeenReached) {
                        if (mListener != null) {
                            if (!loadingNewItems) {
                                loadingNewItems = true
                                mListener!!.onNextPage(++mCurrentPage)
                            }
                        }
                    }
                }
            }
        })
    }

    fun setOnPaginationListener(onPaginationListener: OnPaginationListener?) {
        mListener = onPaginationListener
    }

    interface OnPaginationListener {
        fun onCurrentPage(page: Int)
        fun onNextPage(page: Int)
        fun onFinish()
    }
}